File: Java/TREE/Diameter of a Binary Tree.java
//https://www.geeksforgeeks.org/problems/diameter-of-binary-tree/1
class Solution {
    private int helper(Node root, int[] res) {
        // Base case: if the node is null, return height 0
        if (root == null) {
            return 0;
        }

        // Recursive call to get the height of left and right subtrees
        int leftHeight = helper(root.left, res);
        int rightHeight = helper(root.right, res);

        // Update the diameter (maximum of current diameter and sum of left and right heights)
        res[0] = Math.max(res[0], leftHeight + rightHeight);

        // Return the height of the current node (1 + max of left and right heights)
        return Math.max(leftHeight, rightHeight) + 1;
    }

    // Function to calculate the diameter of the binary tree
    public int diameter(Node root) {
        int[] res = new int[1];  // This will hold the maximum diameter
        helper(root, res);
        return res[0];  // The result is stored in res[0]
    }
}

File: Python/TREE/Diameter of a Binary Tree.py
//https://www.geeksforgeeks.org/problems/diameter-of-binary-tree/1
class Solution:
    def helper(self, root, res):
        # Base case: if the node is null, return 0
        if not root:
            return 0
        
        # Recursive calls for left and right subtrees
        left_height = self.helper(root.left, res)
        right_height = self.helper(root.right, res)
        
        # Update the diameter (max of current diameter and left + right height)
        res[0] = max(res[0], left_height + right_height)
        
        # Return the height of the current node
        return max(left_height, right_height) + 1

    # Function to calculate the diameter of the binary tree
    def diameter(self, root):
        res = [0]  # This holds the result, using list to pass by reference
        self.helper(root, res)
        return res[0] 
